### PR TITLE
Better variable naming for proofs

### PR DESCRIPTION
**Pull request by ciaranm**

### CODE REVIEW

**Comment by ciaranm** [comment]
As an example of what this produces,

```
{
"i0_banana_b0": {"cpvarid":0,"cpvartype":"intvar","name":"banana","power":0,"type":"intvarbit"},
"i0_banana_b1": {"cpvarid":0,"cpvartype":"intvar","name":"banana","power":1,"type":"intvarbit"},
"i0_banana_b2": {"cpvarid":0,"cpvartype":"intvar","name":"banana","power":2,"type":"intvarbit"},
"i0_banana_b3": {"cpvarid":0,"cpvartype":"intvar","name":"banana","power":3,"type":"intvarbit"},
"i0_banana_b4": {"cpvarid":0,"cpvartype":"intvar","name":"banana","power":4,"type":"intvarbit"},
"i0_banana_b5": {"cpvarid":0,"cpvartype":"intvar","name":"banana","power":5,"type":"intvarbit"},
"i0_banana_b6": {"cpvarid":0,"cpvartype":"intvar","name":"banana","power":6,"type":"intvarbit"},
"i1_chocolate_b0": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","power":0,"type":"intvarbit"},
"i1_chocolate_b1": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","power":1,"type":"intvarbit"},
"i1_chocolate_b2": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","power":2,"type":"intvarbit"},
"i1_chocolate_b3": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","power":3,"type":"intvarbit"},
"i1_chocolate_b4": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","power":4,"type":"intvarbit"},
"i1_chocolate_b5": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","power":5,"type":"intvarbit"},
"i1_chocolate_b6": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","power":6,"type":"intvarbit"},
"i2_profit_b0": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":0,"type":"intvarbit"},
"i2_profit_b1": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":1,"type":"intvarbit"},
"i2_profit_b2": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":2,"type":"intvarbit"},
"i2_profit_b3": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":3,"type":"intvarbit"},
"i2_profit_b4": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":4,"type":"intvarbit"},
"i2_profit_b5": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":5,"type":"intvarbit"},
"i2_profit_b6": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":6,"type":"intvarbit"},
"i2_profit_b7": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":7,"type":"intvarbit"},
"i2_profit_b8": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":8,"type":"intvarbit"},
"i2_profit_b9": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":9,"type":"intvarbit"},
"i2_profit_b10": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":10,"type":"intvarbit"},
"i2_profit_b11": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":11,"type":"intvarbit"},
"i2_profit_b12": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":12,"type":"intvarbit"},
"i2_profit_b13": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":13,"type":"intvarbit"},
"i2_profit_b14": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":14,"type":"intvarbit"},
"i2_profit_b15": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":15,"type":"intvarbit"},
"i2_profit_b16": {"cpvarid":2,"cpvartype":"intvar","name":"profit","power":16,"type":"intvarbit"},
"i0_banana_g17": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":">=","type":"condition","value":17},
"i1_chocolate_g0": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":">=","type":"condition","value":0},
"i1_chocolate_g21": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":">=","type":"condition","value":21},
"i0_banana_g0": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":">=","type":"condition","value":0},
"i0_banana_g4": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":">=","type":"condition","value":4},
"i1_chocolate_g14": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":">=","type":"condition","value":14},
"i1_chocolate_g4": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":">=","type":"condition","value":4},
"i2_profit_g2551": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":">=","type":"condition","value":2551},
"i2_profit_g1200": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":">=","type":"condition","value":1200},
"i0_banana_g3": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":">=","type":"condition","value":3},
"i1_chocolate_g2": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":">=","type":"condition","value":2},
"i2_profit_g1651": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":">=","type":"condition","value":1651},
"i2_profit_g1650": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":">=","type":"condition","value":1650},
"i1_chocolate_g1": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":">=","type":"condition","value":1},
"i0_banana_e3": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":"=","type":"condition","value":3},
"i1_chocolate_e1": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":"=","type":"condition","value":1},
"i2_profit_e1650": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":"=","type":"condition","value":1650},
"i1_chocolate_e0": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":"=","type":"condition","value":0},
"i2_profit_g2151": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":">=","type":"condition","value":2151},
"i1_chocolate_g3": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":">=","type":"condition","value":3},
"i0_banana_g2": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":">=","type":"condition","value":2},
"i2_profit_g1700": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":">=","type":"condition","value":1700},
"i2_profit_g1701": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":">=","type":"condition","value":1701},
"i0_banana_e2": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":"=","type":"condition","value":2},
"i1_chocolate_e2": {"cpvarid":1,"cpvartype":"intvar","name":"chocolate","operator":"=","type":"condition","value":2},
"i2_profit_e1700": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":"=","type":"condition","value":1700},
"i2_profit_g1751": {"cpvarid":2,"cpvartype":"intvar","name":"profit","operator":">=","type":"condition","value":1751},
"i0_banana_g1": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":">=","type":"condition","value":1},
"i0_banana_e1": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":"=","type":"condition","value":1},
"i0_banana_e0": {"cpvarid":0,"cpvartype":"intvar","name":"banana","operator":"=","type":"condition","value":0}
}
```

**Comment by mmcilree** [comment]
Looks good to me. 

A few of minor points: 
1. It looks like there's an out-of-place name being output before each line in the JSON for ProofFlags (think [this line](https://github.com/ciaranm/glasgow-constraint-solver/blob/e5d657d0c55e7c0c321310cd679989499dc82392/gcs/innards/proofs/variable_constraints_tracker.cc#L486) is the offender. 

2. I'm still not entirely sold on `XLiteral` being the most intuitive name for the raw PB literals ... although I can't think of anything better so I guess fine for now. 

3. Should intvarnegbit get a "power" attribute too for consistency? Otherwise, an external tool would have to infer it if it wanted to know for some reason.

4. Now that all the user-supplied names get put into the proof, should we do something to either sanitise the names or throw an error on bad names so doing something like `auto x = p.create_integer_variable_vector(8, 0_i, 7_i, "oh no");` doesn't break everything?

These last two are maybe future nice to haves that I could implement at some point if you like; but open to being told it's not a good idea:

5. Could we have a flag in the JSON to distinguish between "real" variables specified by the user and auxiliary CP variables that are introduced automatically by the solver/modelling process? Like `uservar` set to true by default for any variable created defined through `Problem{}`, but possible to override it so that we can distinguish what the original XCSP/CPMpy/MiniZinc vars are without creating an additional mapping. 

6. Another nice to have: what about having a definition_constraint_id attribute that gives the id of the reifying constraints for proof flags? Or maybe this is more data for the constraints mapping file (".conmap"?) that you and Arthur were discussing?

File: gcs/constraints/all_different/gac_all_different.cc
[the remark this answers is not in the document]
Comment (ciaranm) [reply gcs/constraints/all_different/gac_all_different.cc:1]: This probably should have been in an earlier commit... There's a slight weirdness with auto parameters that can lead to strange warnings about "subobject linkage". Basically, auto parameters are templates in disguise, and if you are relying upon a template being instantiated then either you need the implementation available, or for there to be an explicit instantiation somewhere that has the implementation available.

File: gcs/innards/proofs/variable_constraints_tracker.cc [hunk 1]
```diff
@@ -399,13 +360,241 @@ auto VariableConstraintsTracker::need_gevar(SimpleOrProofOnlyIntegerVariableID i
...
+
+    if (_imp->verbose_names) {
+        string name = "f" + to_string(flag.index) + "_" + name_of(flag);
+        _imp->xlits_to_verbose_names.emplace(result, name);
+        _imp->xlits_to_verbose_names.emplace(! result, "~" + name);
+    }
+
+    nlohmann::json data;
+    data["type"] = "proofflag";
+    _imp->variables_map_file << pb_file_string_for(result) << ' ';
```
[the remark this answers is not in the document]
Comment (ciaranm) [reply gcs/innards/proofs/variable_constraints_tracker.cc:486]: Well spotted

**Pull request closed without merging**